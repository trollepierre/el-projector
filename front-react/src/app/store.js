import { withReducer } from '../store/withReducer';
import { apiService, tokenService } from '../services';

const SET_IS_AUTHENTICATED = 'app/SET_IS_AUTHENTICATED';

const setIsAuthenticated = isAuthenticated => dispatch => {
  if(!isAuthenticated){
    tokenService.removeTokens()
  }
  return dispatch({
    type: SET_IS_AUTHENTICATED, payload: { isAuthenticated }
  });
};

const authenticate = password => async dispatch => {
  try {
    const data = {
      email: 'some@example.org',
      name: password
    };
    const tokens = await apiService.post('login', data)
    await tokenService.saveUserTokens(tokens, data)
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    alert('wrong password')
  }
}

const loginSilently = () => async dispatch => {
  try {
    const refreshToken = tokenService.getRefreshToken();
    if(!refreshToken) return;
    const tokens = await apiService.post('login/token', { refreshToken })
    await tokenService.reauthenticate(tokens);
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    await tokenService.removeTokens()
    alert('problem during login silently')
    throw error
  }
}

export const actions = {
  setIsAuthenticated,
  authenticate,
  loginSilently,
};

export const initialState = {
  isAuthenticated: false,
};

export const actionHandlers = {
  [SET_IS_AUTHENTICATED]: (state, action) => {
    const { isAuthenticated } = action.payload;
    return {
      ...state,
      isAuthenticated,
    };
  },
};

export const reducer = withReducer(initialState, actionHandlers);
