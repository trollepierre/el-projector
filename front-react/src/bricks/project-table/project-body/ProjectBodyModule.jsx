import React, { useEffect } from 'react';
import TaskRaw from './task-raw/TaskRaw';
import PropTypes from 'prop-types';

const ProjectBodyModule = ({
    tasks,
    fetchTasks,
  }) => {
  const { data, isLoading, error } = tasks

  useEffect(() => {
    fetchTasks()
  }, [fetchTasks])

  return (
  <tbody className="project-body">
    {isLoading ? (<tr><td>loading</td></tr>) : null}
    {error ? <tr><td>An error has occured</td></tr> : null}
    {(!data) ? (<tr><td>oops not possible?</td></tr>) :
      data.map(task => (<TaskRaw task={task} key={task.id}/>))}
    </tbody>
    );
 };

ProjectBodyModule.displayName = 'ProjectBodyModule';

ProjectBodyModule.propTypes = {
  tasks: PropTypes.exact({
    isLoading: PropTypes.bool,
    error: PropTypes.object,
    data: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        description: PropTypes.string,
        points: PropTypes.number,
        createdDate: PropTypes.string,
        endDate: PropTypes.string,
      })
    ),
  }),
  fetchTasks: PropTypes.func.isRequired,
};

ProjectBodyModule.defaultProps = {
  tasks: {
  isLoading: true,
  error: undefined,
  data: []
  }
};

export default ProjectBodyModule;
