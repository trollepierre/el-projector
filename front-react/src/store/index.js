export { default as withReducer } from './withReducer'
export { default as createStore } from './createStore'
