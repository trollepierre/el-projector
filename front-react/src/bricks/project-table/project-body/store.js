import apiService from '../../../services/api/api-service'
import { withReducer } from '../../../store/withReducer';

export const FETCH_TASKS_STARTED = 'FETCH_TASKS_STARTED';
export const FETCH_TASKS_SUCCEEDED = 'FETCH_TASKS_SUCCEEDED';
export const FETCH_TASKS_FAILED = 'FETCH_TASKS_FAILED';

// this.sortedTasks = apiService.get('tasks').then(tasks => {
//   console.log('tasks');
//   console.log(tasks);
//   return tasks
// })

export const fetchTasks = () => async dispatch => {
  dispatch({ type: FETCH_TASKS_STARTED });

  try {
    // const tasksFixtures = [
    //   {
    //     id: 1,
    //     image_src:
    //       'https://cdn.manomano.fr/martillo-dewalt-dch273n-18v-P-2075566-4162147_1.jpg',
    //     title: 'Demo react Product',
    //     price_currency: '€',
    //     price: '2490.50',
    //   },
    //   {
    //     id: 2,
    //     image_src:
    //       'https://cdn.manomano.com/futbolin-de-acero-60-kg-140x745x875-cm-negro-P-334554-10088120_1.jpg',
    //     title: 'DIY Essential',
    //     price_currency: '€',
    //     price: '240.50',
    //   },
    // ];
    const data = await apiService.get('tasks');
    dispatch({
      type: FETCH_TASKS_SUCCEEDED,
      payload: data,
    });
  } catch (error) {
    dispatch({
      type: FETCH_TASKS_FAILED,
      error,
    });
  }
};

export const initialState = {
  tasks: {
    data: [],
    isLoading: false,
    error: undefined,
  },
};

export const actionHandlers = {
  [FETCH_TASKS_STARTED]: state => {
    return {
      ...state,
      tasks: {
        ...state.tasks,
        error: undefined,
        isLoading: true,
      },
    };
  },
  [FETCH_TASKS_SUCCEEDED]: (state, action) => {
    return {
      ...state,
      tasks: {
        data: action.payload,
        error: undefined,
        isLoading: false,
      },
    };
  },
  [FETCH_TASKS_FAILED]: (state, action) => {
    return {
      ...state,
      tasks: {
        ...state.tasks,
        error: action.error,
        isLoading: false,
      },
    };
  },
};

export const reducer = withReducer(initialState, actionHandlers);

export const tasksSelector = state => state.tasks;

export default reducer;

