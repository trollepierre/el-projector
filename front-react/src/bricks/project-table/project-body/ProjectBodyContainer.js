/* istanbul ignore next */
import { fetchTasks, tasksSelector } from './store';
import { connect } from 'react-redux';
import { ProjectBodyModule } from './ProjectBodyModule';

const mapStateToProps = state => {
  return tasksSelector(state);
};

const mapDispatchToProps = {
  fetchTasks,
};

/* istanbul ignore next */
export default
connect(
  mapStateToProps,
  mapDispatchToProps
)(ProjectBodyModule)
;

