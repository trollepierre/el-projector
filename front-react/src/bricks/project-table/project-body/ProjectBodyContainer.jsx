import React, { useEffect, useReducer } from 'react';
import ProjectBodyModule from './ProjectBodyModule';
import { actions, initialState, reducer, tasksSelector } from './store';
import { useAppContext } from '../../../app/AppContext';

const ProjectBodyContainer = () => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const { loginSilently } = useAppContext()

  useEffect(() => {
    actions.fetchTasks(loginSilently)(dispatch)
  }, [loginSilently]);

  return <ProjectBodyModule tasks={tasksSelector(state)} />;
};

export default ProjectBodyContainer;
