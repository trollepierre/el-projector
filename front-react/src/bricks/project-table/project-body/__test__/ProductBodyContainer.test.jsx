import React, { useReducer } from 'react';
import { shallow } from 'enzyme';
import ProjectBodyContainer from '../ProjectBodyContainer';
import ProjectBodyModule from '../ProjectBodyModule';
import { actions } from '../store';

describe('ProjectBodyContainer', () => {
  it('should match snapshot', () => {
    // When
    const wrapper = shallow(<ProjectBodyContainer/>);

    // Then
    expect(wrapper).toMatchSnapshot();
  });

  describe('events', () => {
    it('should call fetch tasks', () => {
      // Given
      actions.fetchTasks = jest.fn(() => () => {})
      const wrapper = shallow(<ProjectBodyContainer/>);

      // When
      wrapper.find(ProjectBodyModule).props().fetchTasks('loginSilently')

      // Then
      expect(actions.fetchTasks).toHaveBeenCalledWith('loginSilently');
    });
  });
});
