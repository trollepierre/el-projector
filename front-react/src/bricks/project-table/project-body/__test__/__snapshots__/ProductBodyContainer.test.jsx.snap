// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectBodyContainer should match snapshot 1`] = `
<ProjectBodyModule
  fetchTasks={[Function]}
  tasks={
    Object {
      "data": Array [],
      "error": undefined,
      "isLoading": false,
    }
  }
/>
`;
